xarray/core/dataset.py

```diff
@@ -1212,7 +1212,18 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
         coords: dict[Hashable, Variable] = {}
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self._variables[k].dims) <= needed_dims:
+            var_dims = set(self._variables[k].dims)
+            if k in self._indexes:
```

Comment:
I am a little concerned from a performance perspective about introducing a double loop over variables/indexes inside DataArray creation.

I need to think a little harder about what the alternative would be, though...